eclair-core/src/main/scala/fr/acinq/eclair/router/RouteCalculation.scala

```diff
@@ -146,63 +154,68 @@ object RouteCalculation {
+                                localNodeId: PublicKey,
+                                targetNodeId: PublicKey,
+                                amount: MilliSatoshi,
+                                maxFee: MilliSatoshi,
+                                numRoutes: Int,
+                                extraEdges: Set[GraphEdge] = Set.empty,
+                                ignoredEdges: Set[ChannelDesc] = Set.empty,
+                                ignoredVertices: Set[PublicKey] = Set.empty,
+                                routeParams: RouteParams,
+                                currentBlockHeight: Long): Either[RouterException, Seq[Graph.WeightedPath]] = {
+    if (localNodeId == targetNodeId) return Left(CannotRouteToSelf)
+
+    def feeOk(fee: MilliSatoshi): Boolean = fee <= maxFee
 
     def lengthOk(length: Int): Boolean = length <= routeParams.routeMaxLength && length <= ROUTE_MAX_LENGTH
 
     def cltvOk(cltv: CltvExpiryDelta): Boolean = cltv <= routeParams.routeMaxCltv
 
-    val boundaries: RichWeight => Boolean = { weight =>
-      feeOk(weight.cost - amount, amount) && lengthOk(weight.length) && cltvOk(weight.cltv)
-    }
-
-    val foundRoutes = KamonExt.time(Metrics.FindRouteDuration.withTag(Tags.NumberOfRoutes, numRoutes).withTag(Tags.Amount, Tags.amountBucket(amount))) {
-      Graph.yenKshortestPaths(g, localNodeId, targetNodeId, amount, ignoredEdges, ignoredVertices, extraEdges, numRoutes, routeParams.ratios, currentBlockHeight, boundaries).toList
-    }
-    foundRoutes match {
-      case Nil if routeParams.routeMaxLength < ROUTE_MAX_LENGTH => // if not found within the constraints we relax and repeat the search
-        Metrics.RouteLength.withTag(Tags.Amount, Tags.amountBucket(amount)).record(0)
-        return findRoute(g, localNodeId, targetNodeId, amount, numRoutes, extraEdges, ignoredEdges, ignoredVertices, routeParams.copy(routeMaxLength = ROUTE_MAX_LENGTH, routeMaxCltv = DEFAULT_ROUTE_MAX_CLTV), currentBlockHeight)
-      case Nil =>
-        Metrics.RouteLength.withTag(Tags.Amount, Tags.amountBucket(amount)).record(0)
-        throw RouteNotFound
-      case foundRoutes =>
-        val routes = foundRoutes.find(_.path.size == 1) match {
-          case Some(directRoute) => directRoute :: Nil
-          case _ => foundRoutes
-        }
-        // At this point 'routes' cannot be empty
-        val randomizedRoutes = if (routeParams.randomize) Random.shuffle(routes) else routes
-        val route = randomizedRoutes.head.path.map(graphEdgeToHop)
-        Metrics.RouteLength.withTag(Tags.Amount, Tags.amountBucket(amount)).record(route.length)
-        Route(amount, route)
+    val boundaries: RichWeight => Boolean = { weight => feeOk(weight.cost - amount) && lengthOk(weight.length) && cltvOk(weight.cltv) }
+
+    val foundRoutes = Graph.yenKshortestPaths(g, localNodeId, targetNodeId, amount, ignoredEdges, ignoredVertices, extraEdges, numRoutes, routeParams.ratios, currentBlockHeight, boundaries)
```

Comment:
Probably better to type this explicitely, because your implementation later assumes that the collection order is maintained (`++` concatenation).

```suggestion
    val foundRoutes: Seq[WeightedPath] = Graph.yenKshortestPaths(g, localNodeId, targetNodeId, amount, ignoredEdges, ignoredVertices, extraEdges, numRoutes, routeParams.ratios, currentBlockHeight, boundaries)
```